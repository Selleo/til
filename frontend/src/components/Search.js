import React, { useState, useEffect } from 'react'
import { useHistory, useLocation } from 'react-router-dom'
import { useDispatch } from 'react-redux'
import { saveSearchedPosts, saveSearchedQuery } from '../store/actions/actions'
import { useOnRouteLeave } from '../utils/customHooks/useOnRouteLeave'
import { ReactComponent as SearchIcon } from '../assets/icons/search.svg'
import { toast } from 'react-toastify'
import 'react-toastify/dist/ReactToastify.css'

const Search = () => {
  const [input, setInput] = useState('')
  const [timeoutID, setTimeoutID] = useState(null)
  const [disableInput, setDisableInput] = useState(false)
  const dispatch = useDispatch()
  const history = useHistory()
  const location = useLocation()
  const hasLeavedRoute = useOnRouteLeave('/search')

  useEffect(() => {
    if (hasLeavedRoute) {
      dispatch(saveSearchedQuery(''))
      setInput('')
    }
  }, [hasLeavedRoute, dispatch])

  useEffect(() => {
    if (
      location.pathname.includes('add') ||
      location.pathname.includes('edit')
    ) {
      setDisableInput(true)
    } else {
      setDisableInput(false)
    }
  }, [location])

  const handleInput = event => {
    const targetValue = event.target.value

    setInput(targetValue)
    clearTimeout(timeoutID)

    dispatch(saveSearchedQuery(targetValue))

    if (targetValue) {
      const timeout = setTimeout(() => {
        history.push('/search')
        dispatch(saveSearchedPosts(targetValue))
      }, 300)
      setTimeoutID(timeout)
    }
  }
  const notify = () => toast("Can't search while post is creating/editing")

  return (
    <div className="search-box" onClick={disableInput && notify}>
      <input
        className="search-box__input"
        type="text"
        placeholder="Search"
        value={input}
        disabled={disableInput}
        onChange={handleInput}
      />

      <SearchIcon className="search-box__icon" />
    </div>
  )
}

export default Search
