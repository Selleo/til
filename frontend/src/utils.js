import { useLocation } from "react-router-dom";
const localStorageKey = "til_token";

// FETCH POSTS
export const fetchData = async url => {
  let optionsToken = null;

  if (getToken()) {
    optionsToken = {
      headers: {
        Authorization: `Bearer ${getToken()}`,
        "Content-Type": "application/json"
      }
    };
  }

  const response = await fetch(url, optionsToken);
  const data = await response.json();

  return data;
};

export const fetchSinglePost = async (url, id) => {
  let optionsToken = null;

  if (getToken()) {
    optionsToken = {
      headers: {
        Authorization: `Bearer ${getToken()}`,
        "Content-Type": "application/json"
      }
    };
  }

  const response = await fetch(url + id, optionsToken);
  const post = response.json();

  return post;
};

export const fetchReviewPost = async url => {
  const response = await fetch(url, {
    headers: {
      Authorization: `Bearer ${getToken()}`,
      "Content-Type": "application/json"
    }
  });

  return response.json();
};

// APPROVE POST
export const approvePost = async url => {
  const response = await fetch(url, {
    method: "PUT",
    headers: {
      Authorization: `Bearer ${getToken()}`,
      "Content-Type": "application/json"
    }
  });

  return response.ok;
};

// ADD POST
export const postData = async (url, data) => {
  const response = await fetch(url, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${getToken()}`,
      "Content-Type": "application/json"
    },
    body: data
  });

  return response.ok;
};

// UPDATE POST
export const updateData = async (url, data) => {
  const response = await fetch(url, {
    method: "PATCH",
    headers: {
      Authorization: `Bearer ${getToken()}`,
      "Content-Type": "application/json"
    },
    body: data
  });

  return response.ok;
};

// DELETE POST
export const deleteData = async (url, data) => {
  const response = await fetch(url, {
    method: "DELETE",
    headers: {
      Authorization: `Bearer ${getToken()}`,
      "Content-Type": "application/json"
    }
  });

  return response.ok;
};

// FETCH USERS POSTS
export const fetchUserPosts = async (url, id) => {
  try {
    const response = await fetch(url + id);
    const user = await response.json();
    const userPosts = user.posts;

    return userPosts;
  } catch (error) {
    console.warn(error);
  }
};

// LIKES
export const handleLike = async (postId, method) => {
  const response = await fetch(`api/posts/${postId}/likes`, {
    method: `${method}`,
    headers: {
      Authorization: `Bearer ${getToken()}`,
      "Content-Type": "application/json"
    }
  });

  return response.ok;
};

export const checkHasLiked = (likes, userId) => {
  const likesArray = likes.map(like => like.user_uuid);

  return likesArray.includes(userId);
};

// AUTHENTICATION
export const isAuthenticated = () => {
  const token = getToken();
  if (token) return true;

  return false;
};

// TOKENS
export const deleteToken = () => {
  window.localStorage.removeItem(localStorageKey);
};

export const getToken = () => {
  return window.localStorage.getItem(localStorageKey);
};

// URLS
export const getCurrentURL = () => {
  return document.URL;
};

export const useQuery = () => {
  const queryString = useLocation().search;

  return new URLSearchParams(queryString);
};

// FETCH USER
export const fetchUser = async url => {
  const response = await fetch(url, {
    method: "GET",
    headers: {
      Authorization: `Bearer ${getToken()}`,
      "Content-Type": "application/json"
    }
  });

  return response.json();
};

// HELPERS
export const convertToSelectOptions = data => {
  return data.map(el => ({ value: el.id, label: el.name }));
};

// REACTIONS
export const handleReaction = async (postId, method, reaction) => {
  let ending = "";

  if (method === "DELETE") {
    ending = `/${reaction}`;
  }

  const response = await fetch(
    `api/posts/${postId}/reactions${ending}`,
    {
      method: `${method}`,
      headers: {
        Authorization: `Bearer ${getToken()}`,
        "Content-Type": "application/json"
      },
      body: JSON.stringify({ type: reaction })
    }
  );

  return response.ok;
};

export const convertReactions = reactions => {
  const reactionsType = ["like", "funny", "love", "surprised"];

  let filteredReactions = [];

  reactionsType.forEach(reaction => {
    const result = reactions.filter(el => el.type === reaction);
    const reactionObj = { type: reaction, whoReacted: result };
    filteredReactions.push(reactionObj);
  });

  return filteredReactions;
};

export const checkHasReacted = (reaction, userId) => {
  const reactionArray = reaction.map(reaction => reaction.user_uuid);

  return reactionArray.includes(userId);
};
