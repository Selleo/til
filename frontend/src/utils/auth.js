const localStorageKey = "til_token";

// TOKENS
export const deleteToken = () => {
  window.localStorage.removeItem(localStorageKey);
};

export const getToken = () => {
  return window.localStorage.getItem(localStorageKey);
};

// AUTHENTICATION
export const isAuthenticated = () => {
  const token = getToken();
  if (token) {
    return true;
  }

  return false;
};
