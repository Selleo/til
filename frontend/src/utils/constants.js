export const statusType = {
  idle: 'idle',
  loading: 'loading',
  fetched: 'fetched',
}
