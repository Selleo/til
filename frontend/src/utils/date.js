const monthNames = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December',
]

export const getDate = () => {
  const today = new Date()
  let minutes = today.getMinutes()
  if (minutes < 10) {
    minutes = `0${today.getMinutes()}`
  }
  const time = `${today.getHours()}:${minutes}`
  const month = monthNames[today.getMonth() - 1].slice(0, 3)
  const date = `${month} ${today.getDate()} | ${time}`

  return date
}
