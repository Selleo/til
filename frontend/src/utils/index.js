export {
  request,
  fetchData,
  fetchSinglePost,
  fetchUserPosts
} from "./posts/CRUDposts";

export { approvePost } from "./reviewPost/approvePost";
export { fetchReviewPost } from "./reviewPost/fetchReviewPost";

export { getCurrentURL, useQuery } from "./url/getURLparams";

export { handleReaction } from "./posts/handleReactions";

export { fetchUser } from "./dataFetching/fetchUser";

export {
  convertToSelectOptions,
  convertReactions,
  checkHasReacted
} from "./array/helpers";

export { isAuthenticated, deleteToken } from "./auth";
