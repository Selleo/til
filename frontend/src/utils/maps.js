export const categoryStatementMap = new Map([
  [
    'rails',
    {
      textFirst: 'Need help with your Ruby on Rails project?',
      textSecond: 'Visit our website to get more insights.',
    },
  ],
  [
    'ruby',
    {
      textFirst: 'Need help with your Ruby on Rails project?',
      textSecond: 'Visit our website to get more insights.',
    },
  ],
  [
    'javascript',
    {
      textFirst: 'Interested in developing a JavaScript project?',
      textSecond: 'Get more details from our experts.',
    },
  ],
  [
    'general',
    {
      textFirst: 'Need key insights on software development?',
      textSecond: 'Go to our main page to get more tips.',
    },
  ],
  [
    'react',
    {
      textFirst: 'Need help with React.js?',
      textSecond: 'Click for more tips and tricks.',
    },
  ],
  [
    'reactnative',
    {
      textFirst: 'Interested in building a powerful mobile solution?',
      textSecond: 'Get insights from experienced developers.',
    },
  ],
  [
    'android',
    {
      textFirst: 'Interested in building a powerful mobile solution?',
      textSecond: 'Get insights from experienced developers.',
    },
  ],
  [
    'css',
    {
      textFirst: 'Looking for insights on UX/UI development?',
      textSecond: 'Visit our page to learn more about it.',
    },
  ],
  [
    'html',
    {
      textFirst: 'Looking for insights on UX/UI development?',
      textSecond: 'Visit our page to learn more about it.',
    },
  ],
  [
    'git',
    {
      textFirst: 'Need key insights on software development?',
      textSecond: 'Go to our main page to get more tips.',
    },
  ],
  [
    'chrome',
    {
      textFirst: 'Need key insights on software development?',
      textSecond: 'Go to our main page to get more tips.',
    },
  ],
  [
    'linux',
    {
      textFirst: 'Need key insights on software development?',
      textSecond: 'Go to our main page to get more tips.',
    },
  ],
  [
    'devops',
    {
      textFirst: 'Transform your DevOps expertise today!',
      textSecond: 'Get more tips from our tech leaders.',
    },
  ],
  [
    'commandline',
    {
      textFirst: 'Transform your DevOps expertise today!',
      textSecond: 'Get more tips from our tech leaders.',
    },
  ],
  [
    'terraform',
    {
      textFirst: 'Transform your DevOps expertise today!',
      textSecond: 'Get more tips from our tech leaders.',
    },
  ],
  [
    'aws',
    {
      textFirst: 'Transform your DevOps expertise today!',
      textSecond: 'Get more tips from our tech leaders.',
    },
  ],
  [
    'crystal',
    {
      textFirst: 'Transform your DevOps expertise today!',
      textSecond: 'Get more tips from our tech leaders.',
    },
  ],
  [
    'sql',
    {
      textFirst: 'Transform your DevOps expertise today!',
      textSecond: 'Get more tips from our tech leaders.',
    },
  ],
  [
    'docker',
    {
      textFirst: 'Transform your DevOps expertise today!',
      textSecond: 'Get more tips from our tech leaders.',
    },
  ],
  [
    'heroku',
    {
      textFirst: 'Transform your DevOps expertise today!',
      textSecond: 'Get more tips from our tech leaders.',
    },
  ],
  [
    'cicd',
    {
      textFirst: 'Transform your DevOps expertise today!',
      textSecond: 'Get more tips from our tech leaders.',
    },
  ],
  [
    'vault',
    {
      textFirst: 'Transform your DevOps expertise today!',
      textSecond: 'Get more tips from our tech leaders.',
    },
  ],
  [
    'elixir',
    {
      textFirst: 'Looking for Elixir expertise?',
      textSecond: 'Read more from our experts.',
    },
  ],
  [
    'phoenix',
    {
      textFirst: 'Looking for Elixir expertise?',
      textSecond: 'Read more from our experts.',
    },
  ],
  [
    'nodejs',
    {
      textFirst: 'Need help with Node.js development?',
      textSecond: 'Get help from our experienced teams',
    },
  ],
  [
    'ember',
    {
      textFirst: 'Looking for tips from experienced Ember developers?',
      textSecond: 'Visit our website for more information.',
    },
  ],
  [
    'qa',
    {
      textFirst: 'In need of help with the quality of your products?',
      textSecond: 'Visit our page to learn more about it.',
    },
  ],
])
