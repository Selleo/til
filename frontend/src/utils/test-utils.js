import React from "react";
import { Provider } from "react-redux";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import configureMockStore from "redux-mock-store";
import { initialState as defaultState } from "../dummyData";

export const WrapperWithStoreAndRouter = ({
  children,
  initialState = defaultState,
}) => {
  const history = createMemoryHistory();
  const mockStore = configureMockStore([]);
  const store = mockStore(initialState);

  return (
    <Provider store={store}>
      <Router history={history}>{children}</Router>
    </Provider>
  );
};
export const WrapperWithStore = ({ children, initialState = defaultState }) => {
  const mockStore = configureMockStore([]);
  const store = mockStore(initialState);

  return <Provider store={store}>{children}</Provider>;
};

export const WrapperWithRouter = ({ children }) => {
  const history = createMemoryHistory();

  return <Router history={history}>{children}</Router>;
};
