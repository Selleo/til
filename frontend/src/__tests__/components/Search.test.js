import React from "react";
import { render, fireEvent, cleanup } from "@testing-library/react";
import Search from "../../components/Search";
import { WrapperWithStoreAndRouter } from "../../utils/test-utils";

describe("<Search/>", () => {
  let component;

  beforeEach(() => {
    component = render(
      <WrapperWithStoreAndRouter>
        <Search />
      </WrapperWithStoreAndRouter>
    );
  });

  afterEach(cleanup);

  it("renders correctly", () => {
    const { container } = component;

    expect(container).toMatchSnapshot();
  });

  it("shows search query in input", () => {
    const { getByPlaceholderText } = component;
    const input = getByPlaceholderText(/search/i);

    fireEvent.change(input, { target: { value: "react" } });
    expect(input.value).toBe("react");
  });
});
