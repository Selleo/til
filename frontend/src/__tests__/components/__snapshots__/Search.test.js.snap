// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search/> renders correctly 1`] = `
<div>
  <div>
    <input
      placeholder="search"
      type="text"
      value=""
    />
  </div>
</div>
`;
